### PR TITLE
log: print unexpected version warning in validation log category

### LINKED ISSUE #19603: Show the "<n> of the last 100 blocks have unexpected version" warning only when running -debug=validation?
[Google search results](https://www.google.com/search?q=of+the+last+100+blocks+have+unexpected+version) suggest that our users are more confused than helped by the `<n> of the last 100 blocks
have unexpected version` warning we're printing.

As developers we know the unfortunate reason behind this warning and how "unexpected" should be interpreted in this context, and thus why it is safe to disregard this warning. However, I don't think it reasonable to expect our users to know the historical context here.

In order to not desensitize our users to potentially critical "real" warnings, would it make sense to move this message to the `-debug=validation` log category?

### PR DESCRIPTION
**Pull request by n-thumann**
Fixes #19603: As suggested by practicalswift, instead of always printing `<n> of the last 100 blocks have unexpected version` as a warning appended to UpdateTip, it is now printed in the validation log category and therefore only visible with `-debug=validation` enabled.

Before:
`2020-09-06T15:56:00Z UpdateTip: new best=00000000000000000001b2872e107a98b57913120e5c6c87ce2715a34c40adf8 height=646969 version=0x20400000 log2_work=92.261571 tx=565651941 date='2020-09-06T10:35:36Z' progress=0.999888 cache=32.2MiB(237417txo) warning='72 of last 100 blocks have unexpected version'`
After:
`2020-09-06T16:31:26Z UpdateTip: new best=0000000000000000000b3bd786dc42745dd7be4a8c695500a04518cb9e2f4dc1 height=646971 version=0x20000000 log2_work=92.261607 tx=565655901 date='2020-09-06T10:57:19Z' progress=0.999883 cache=3.8MiB(27550txo)`
`2020-09-06T16:31:26Z 71 of last 100 blocks have unexpected version`

Ran unit & functional tests, confirmed that the warning is now only printed when validation category is enabled.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2465,15 +2465,16 @@ static void UpdateTip(CTxMemPool& mempool, const CBlockIndex* pindexNew, const C
...
     }
     LogPrintf("%s: new best=%s height=%d version=0x%08x log2_work=%f tx=%lu date='%s' progress=%f cache=%.1fMiB(%utxo)%s\n", __func__,
       pindexNew->GetBlockHash().ToString(), pindexNew->nHeight, pindexNew->nVersion,
       log(pindexNew->nChainWork.getdouble())/log(2.0), (unsigned long)pindexNew->nChainTx,
       FormatISO8601DateTime(pindexNew->GetBlockTime()),
       GuessVerificationProgress(chainParams.TxData(), pindexNew), ::ChainstateActive().CoinsTip().DynamicMemoryUsage() * (1.0 / (1<<20)), ::ChainstateActive().CoinsTip().GetCacheSize(),
       !warning_messages.empty() ? strprintf(" warning='%s'", warning_messages.original) : "");
+
+    if (nUpgraded > 0)
+        LogPrint(BCLog::VALIDATION, "%d of last 100 blocks have unexpected version\n", nUpgraded);
```
Comment (hebasto) [inline src/validation.cpp:2477]: Please follow code style for new code:
```suggestion
    if (nUpgraded > 0) {
        LogPrint(BCLog::VALIDATION, "%d of last 100 blocks have unexpected version\n", nUpgraded);
    }
```

**Review by hebasto — COMMENTED** [review]
Concept ACK.

It seems out of scope of this PR, but I think the name `nUpgraded` is a bit misleading now.

File: src/validation.cpp (hunk 1 above)
Comment (n-thumann) [reply src/validation.cpp]: thanks, done :)

**Comment by pox** [comment]
A blessed change!

Tested on Ubuntu by observing logs with default logging:
```
2020-09-06T19:01:17Z UpdateTip: new best=0000000000000000000c35da7da08dd7b1521a3ffcbe9efb5e8eedde53d7896b height=647000 version=0x20000000 log2_work=92.262139 tx=565693850 date='2020-09-06T13:53:09Z' progress=0.999892 cache=1.7MiB(12481txo)
2020-09-06T19:02:51Z UpdateTip: new best=0000000000000000000c5a85e128e2a1b6ef8df242c68e3284a5ac29ecaa159a height=647001 version=0x2000e000 log2_work=92.262157 tx=565694885 date='2020-09-06T13:53:38Z' progress=0.999892 cache=3.0MiB(22258txo)
2020-09-06T19:03:27Z UpdateTip: new best=00000000000000000008e73bdaca4a47cd3926d1dcc728238cb7de2bddd014c0 height=647002 version=0x20c00000 log2_work=92.262175 tx=565696981 date='2020-09-06T14:05:46Z' progress=0.999896 cache=4.3MiB(32499txo)
2020-09-06T19:03:57Z UpdateTip: new best=0000000000000000000d2b11cc472ca88e0e84fc9fdf210eff1378ab6f6191fd height=647003 version=0x37ffe000 log2_work=92.262194 tx=565699661 date='2020-09-06T14:22:45Z' progress=0.999901 cache=5.9MiB(43169txo)
2020-09-06T19:04:21Z UpdateTip: new best=0000000000000000000b39032cc59ffe754832bfb7b1bef4dee75a5bef0be502 height=647004 version=0x20000000 log2_work=92.262212 tx=565702102 date='2020-09-06T14:34:51Z' progress=0.999906 cache=7.2MiB(53849txo)
```

And with `-debug=validation`:

```
2020-09-06T19:08:55Z BlockChecked: block hash=0000000000000000000952b5504198cd2cb4d94041f09e3c81ae42ed3bb691c0 state=Valid
2020-09-06T19:08:55Z UpdateTip: new best=0000000000000000000952b5504198cd2cb4d94041f09e3c81ae42ed3bb691c0 height=647025 version=0x20400000 log2_work=92.262597 tx=565750210 date='2020-09-06T18:26:24Z' progress=0.999985 cache=26.4MiB(193591txo)
2020-09-06T19:08:55Z 63 of last 100 blocks have unexpected version
2020-09-06T19:08:55Z Enqueuing BlockConnected: block hash=0000000000000000000952b5504198cd2cb4d94041f09e3c81ae42ed3bb691c0 block height=647025
2020-09-06T19:08:55Z Enqueuing UpdatedBlockTip: new block hash=0000000000000000000952b5504198cd2cb4d94041f09e3c81ae42ed3bb691c0 fork block hash=0000000000000000000d235796b5fbc3df24499d78fcfa4e2ac77173d413de15 (in IBD=false)
```

**Comment by practicalswift** [comment]
Concept ACK

Thanks a lot for addressing this: this is causing a lot of unnecessary confusion for our users.

Our users are not helped by permanent warnings that are not actionable.

**Comment by practicalswift** [comment]
Tested ACK 7edc0689fad41cfee2830b89dc8331d19747eb5f

**Review by hebasto — APPROVED** [review]
ACK 7edc0689fad41cfee2830b89dc8331d19747eb5f, tested on Linux Mint 20 (x86_64) with and without `-debug=validation` command line option.

I think it is ok to do not translate log messages. `bitcoinstrings.cpp` will be updated automaGically.

**Comment by laanwj** [comment]
> Our users are not helped by permanent warnings that are not actionable.

I've tried to make this point about this log message, but from what I remember I got a lot of pushback last time.
But I still agree.
ACK 7edc0689fad41cfee2830b89dc8331d19747eb5f

**Review by maflcko — COMMENTED** [review]
ACK 7edc0689fad41cfee2830b89dc8331d19747eb5f

If a softfork is deployed without a BIP9 versionbits compatible warning, it is impossible to write code in advance to detect that scenario with full accuracy. A softfork deployment without a versionbits compatible warning (and especially one that would *require* most users to upgrade on a short notice) is reckless at best. Putting code into Bitcoin Core to encourage short notice "forced" upgrades seems counter-productive.

Also, the warning is disabled for users that sync up their nodes intermittently (e.g. once a day or once a week).

Finally, for the remaining well prepared users that:
* Run a node 24/7
* Have -debug or -debug=validation enabled
* Read the debug.log
I'd be surprised if this warning had any more impact on their decision to upgrade to a newer version or look out for new versions than the already existing release announcements and our well-known (?) EOL policy.

So I'd say to simply remove this warning.

File: src/validation.cpp [hunk 2]
```diff
@@ -2441,9 +2441,9 @@ static void UpdateTip(CTxMemPool& mempool, const CBlockIndex* pindexNew, const C
     }
 
     bilingual_str warning_messages;
+    int nUpgraded = 0;
```
[the remark this answers is not in the document]
Comment (n-thumann) [reply src/validation.cpp:2444]: good point, done

**Comment by practicalswift** [comment]
ACK 62dba9628d2532dca0c44934600f5aac3b21e275 -- only change since last ACK is `s/nUpgraded/num_unexpected_version/`

**Review by hebasto — APPROVED** [review]
re-ACK 62dba9628d2532dca0c44934600f5aac3b21e275, https://github.com/bitcoin/bitcoin/pull/19898#pullrequestreview-483158708 is resolved now.

**Review by theStack — APPROVED** [review]
ACK 62dba9628d2532dca0c44934600f5aac3b21e275

**Comment by practicalswift** [comment]
Is this one ready for merge? :)

I'm counting 4 ACKs (practicalswift, MarcoFalke, hebasto, theStack) and one stale ACK (laanwj).

**Review by jonatack — COMMENTED** [review]
Posthumous ACK 62dba9628d2532dca0c44934600f5aac3b21e275

**Comment by bolatovumar** [comment]
Thanks for this change! That message is really annoying.

**Comment by iemwill** [comment]
Line 2476 is a good one.

**Pull request merged**